src/body/incoming.rs

```diff
@@ -201,7 +201,16 @@ impl Body for Incoming {
                             ping.record_data(bytes.len());
                             return Poll::Ready(Some(Ok(Frame::data(bytes))));
                         }
-                        Some(Err(e)) => return Poll::Ready(Some(Err(crate::Error::new_body(e)))),
+                        Some(Err(e)) => {
+                            return match e.reason() {
+                                // These reasons should cause stop of body reading, but don't fail it.
+                                // The same logic as for `Read for H2Upgraded` is applied here.
```

Comment:
I'm referring to this line here: https://github.com/hyperium/hyper/blob/a45d5d5a04369f93334fc893875d8d1a49054e04/src/proto/h2/mod.rs#L296

Both of them call the same operation (`self.recv_stream.poll_data(cx)`) and seems that logic should be the same. But we can make it more specific.

Reply:
Btw, an additional example from another lang & lib where it was also solved by avoiding error if response has been read: https://github.com/square/okhttp/pull/6295